File: sdk/openrouter.py
"""
openrouter.py — drop-in OpenRouter client with automatic token tracking.

Usage:
    from openrouter import OpenRouterClient

    client = OpenRouterClient(api_key="sk-or-...", tracker_url="http://localhost:4242")
    reply = client.chat("openai/gpt-4o-mini", [{"role": "user", "content": "hello"}])
    print(reply["choices"][0]["message"]["content"])
"""

import time
import json
import requests
from datetime import datetime, timezone


class OpenRouterClient:
    def __init__(self, api_key: str, tracker_url: str = "http://localhost:4242", app_name: str = "my-app"):
        if not api_key:
            raise ValueError("api_key is required")
        self.api_key = api_key
        self.tracker_url = tracker_url
        self.app_name = app_name
        self.base_url = "https://openrouter.ai/api/v1"

    def chat(self, model: str, messages: list, **kwargs) -> dict:
        started_at = time.time()

        payload = {"model": model, "messages": messages, **kwargs}
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.api_key}",
            "HTTP-Referer": kwargs.pop("referer", "http://localhost"),
            "X-Title": self.app_name,
        }

        resp = requests.post(
            f"{self.base_url}/chat/completions",
            headers=headers,
            json=payload,
            timeout=60,
        )

        latency_ms = int((time.time() - started_at) * 1000)

        try:
            data = resp.json()
        except Exception:
            data = {}

        if not resp.ok:
            err = data.get("error", {}).get("message") or f"HTTP {resp.status_code}"
            self._report({"model": model, "error": err, "latency_ms": latency_ms})
            raise RuntimeError(f"OpenRouter error: {err}")

        usage = data.get("usage", {})
        last_msg = messages[-1].get("content", "") if messages else ""

        event = {
            "ts": datetime.now(timezone.utc).isoformat(),
            "app": self.app_name,
            "model": model,
            "prompt_tokens": usage.get("prompt_tokens"),
            "completion_tokens": usage.get("completion_tokens"),
            "total_tokens": usage.get("total_tokens"),
            "latency_ms": latency_ms,
            "prompt_preview": str(last_msg)[:120],
            "finish_reason": (data.get("choices") or [{}])[0].get("finish_reason", "unknown"),
            "error": None,
        }

        self._report(event)
        return data

    def _report(self, event: dict):
        try:
            requests.post(
                f"{self.tracker_url}/track",
                json=event,
                timeout=2,
            )
        except Exception:
            pass  # tracker offline — don't crash your app

File: sdk/openrouter.js
/**
 * openrouter.js — drop-in OpenRouter client with automatic token tracking.
 * 
 * Usage:
 *   const { OpenRouterClient } = require('./openrouter');
 *   const client = new OpenRouterClient({ apiKey: 'sk-or-...', trackerUrl: 'http://localhost:4242' });
 *   const reply = await client.chat('openai/gpt-4o-mini', [{ role: 'user', content: 'hello' }]);
 */

class OpenRouterClient {
  constructor({ apiKey, trackerUrl = 'http://localhost:4242', appName = 'my-app' }) {
    if (!apiKey) throw new Error('apiKey is required');
    this.apiKey = apiKey;
    this.trackerUrl = trackerUrl;
    this.appName = appName;
    this.baseUrl = 'https://openrouter.ai/api/v1';
  }

  async chat(model, messages, options = {}) {
    const startedAt = Date.now();

    const body = { model, messages, ...options };

    const res = await fetch(`${this.baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
        'HTTP-Referer': options.referer || 'http://localhost',
        'X-Title': this.appName,
      },
      body: JSON.stringify(body),
    });

    const data = await res.json();

    if (!res.ok) {
      const err = data?.error?.message || `HTTP ${res.status}`;
      await this._report({ model, error: err, latency_ms: Date.now() - startedAt });
      throw new Error(`OpenRouter error: ${err}`);
    }

    const usage = data.usage || {};
    const event = {
      ts: new Date().toISOString(),
      app: this.appName,
      model,
      prompt_tokens: usage.prompt_tokens ?? null,
      completion_tokens: usage.completion_tokens ?? null,
      total_tokens: usage.total_tokens ?? null,
      latency_ms: Date.now() - startedAt,
      prompt_preview: messages.at(-1)?.content?.slice(0, 120) ?? '',
      finish_reason: data.choices?.[0]?.finish_reason ?? 'unknown',
      error: null,
    };

    await this._report(event);
    return data;
  }

  async _report(event) {
    try {
      await fetch(`${this.trackerUrl}/track`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(event),
      });
    } catch {
      // tracker offline — don't crash your app
    }
  }
}

// Also export a convenience wrapper that mirrors the OpenAI SDK style
class OpenRouterChat {
  constructor(clientOptions) {
    this.client = new OpenRouterClient(clientOptions);
  }

  get completions() {
    return {
      create: ({ model, messages, ...opts }) =>
        this.client.chat(model, messages, opts),
    };
  }
}

module.exports = { OpenRouterClient, OpenRouterChat };
